src/components/CalculateClockOptions.ts

```diff
@@ -0,0 +1,161 @@
+import { DataFrame, Field, PanelData, PanelProps } from '@grafana/data';
+import moment, { Moment, MomentInput } from 'moment-timezone';
+import {
+  ClockMode,
+  ClockOptions,
+  ClockSource,
+  CountdownQueryCalculation,
+  CountupQueryCalculation,
+  DescriptionSource,
+  QueryCalculation,
+} from 'types';
+import { getMoment } from 'utils';
+
+type QueryRow = {
+  time: Moment;
+  description: string;
+};
+
+export function CalculateClockOptions({
+  options,
+  timezone,
+  data,
+  replaceVariables,
+  now,
+}: {
+  options: ClockOptions;
+  timezone: ClockOptions['timezone'];
+  data: PanelData;
+  replaceVariables: PanelProps['replaceVariables'];
+  now: Moment;
+}): [Moment, string, string | null] {
+  let descriptionNoValueText = options.descriptionSettings.noValueText ?? '';
+  let descriptionText =
+    options.descriptionSettings.source === DescriptionSource.query
+      ? descriptionNoValueText
+      : options.descriptionSettings.descriptionText;
+  if (options.mode !== ClockMode.countdown && options.mode !== ClockMode.countup) {
+    return [now, descriptionText, null];
+  }
+
+  let userInputTime: string | undefined = '';
+  let clockSettings: ClockOptions['countdownSettings'] | ClockOptions['countupSettings'];
+  switch (options.mode) {
+    case ClockMode.countdown:
+      userInputTime = options.countdownSettings.endCountdownTime || '';
+      clockSettings = options.countdownSettings;
+      break;
+    case ClockMode.countup:
+      userInputTime = options.countupSettings.beginCountupTime || '';
+      clockSettings = options.countupSettings;
+      break;
+  }
+
+  let clockNoValueText = clockSettings.noValueText ?? '';
+  let clockInvalidValueText = clockSettings.invalidValueText ?? '';
+  let targetTime: Moment | undefined = undefined;
+  switch (clockSettings.source) {
+    case ClockSource.input:
+      if (!userInputTime) {
+        return [now, descriptionText, clockNoValueText];
+      }
+      targetTime = moment(replaceVariables(userInputTime)).utcOffset(getMoment(timezone).format('Z'), true);
+      break;
+    case ClockSource.query:
+      const isDataReady = data.state === 'Done' && data.series.length !== 0;
+      const isQueryFieldSelected = clockSettings.queryField && clockSettings.queryField.length !== 0;
+      const isDataSeriesPopulated = data.series && data.series[0].fields && data.series[0].fields.length !== 0;
```

Comment:
it is possible that we have several series data but some of them came empty so having no fields in series[0] doesn't mean the other don't have data.

Reply:
Ok, I will remove the `isDataSeriesPopulated` condition as empty fields would be catched by the `timeField` not finding anything.